src/net/download.rs

```diff
@@ -0,0 +1,1904 @@
+                },
+                Some(url::Host::Ipv4(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V4(addr), port)]));
+                }
+                Some(url::Host::Ipv6(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V6(addr), port)]));
+                },
+                None => {
+                    warn!("Unsupported URL {:?}", &url_str);
+                }
+            }
+        }
+
+        self.state = BlockDownloaderState::DNSLookupFinish;
+        Ok(())
+    }
+
+    pub fn dns_lookups_try_finish(&mut self, dns_client: &mut DNSClient) -> Result<bool, net_error> {
+        dns_client.try_recv()?;
+
+        let mut inflight = 0;
+        for (url_str, request) in self.parsed_urls.iter() {
+            match dns_client.poll_lookup(&request.host, request.port) {
+                Ok(Some(query_result)) => {
+                    if let Some(dns_result) = self.dns_lookups.get_mut(url_str) {
+                        // solicited
+                        match query_result.result {
+                            Ok(addrs) => {
+                                *dns_result = Some(addrs);
+                            },
+                            Err(msg) => {
+                                warn!("DNS failed to look up {:?}: {}", &url_str, msg);
+                            }
+                        }
+                    }
+                },
+                Ok(None) => {
+                    inflight += 1;
+                }
+                Err(e) => {
+                    warn!("DNS lookup failed on {:?}: {:?}", url_str, &e);
+                }
+            }
+        }
+
+        if inflight == 0 {
+            // done with DNS
+            dns_client.clear_all_requests();
+            self.state = BlockDownloaderState::GetBlocksBegin;
+        }
+
+        Ok(inflight == 0)
+    }
+
+    pub fn getblocks_begin(&mut self, mut requests: HashMap<BlockRequestKey, usize>) -> () {
+        assert_eq!(self.state, BlockDownloaderState::GetBlocksBegin);
+
+        // don't touch blocks-to-try -- that's managed by the peer network directly.
+        self.getblock_requests.clear();
+        for (key, event_id) in requests.drain() {
+            self.getblock_requests.insert(key, event_id);
+        }
+
+        self.state = BlockDownloaderState::GetBlocksFinish;
+    }
+
+    /// Finish fetching blocks.  Return true once all reply handles have been fulfilled (either
+    /// with data, or with an error).
+    /// Store blocks as we get them.
+    pub fn getblocks_try_finish(&mut self, http: &mut HttpPeer) -> Result<bool, net_error> {
+        assert_eq!(self.state, BlockDownloaderState::GetBlocksFinish);
+
+        // requests that are still pending
+        let mut pending_block_requests = HashMap::new();
+
+        for (block_key, event_id) in self.getblock_requests.drain() {
+            let rh_block_key = block_key.clone();
```

Comment:
`rh_block_key` doesn't seem like a necessary clone

Reply:
Yes, it is.  It's a hold-over from earlier code.  Will remove.

Reply:
Fixed in 85af7fa